src/libfetchers/git-accessor.cc

```diff
@@ -0,0 +1,422 @@
+{
+    git_oid oid;
+    if (git_oid_fromstr(&oid, hash.gitRev().c_str()))
+        throw Error("cannot convert '%s' to a Git OID", hash.gitRev());
+    return oid;
+}
+
+struct GitInputAccessor : InputAccessor
+{
+    Repository repo;
+    Tree root;
+
+    GitInputAccessor(Repository && repo_, const Hash & rev)
+        : repo(std::move(repo_))
+    {
+        auto oid = hashToOID(rev);
+
+        git_object * obj = nullptr;
+        if (git_object_lookup(&obj, repo.get(), &oid, GIT_OBJECT_ANY)) {
+            auto err = git_error_last();
+            throw Error("getting Git object '%s': %s", rev.gitRev(), err->message);
+        }
+
+        if (git_object_peel((git_object * *) &root, obj, GIT_OBJECT_TREE)) {
+            auto err = git_error_last();
+            throw Error("peeling Git object '%s': %s", rev.gitRev(), err->message);
+        }
+    }
+
+    std::string readFile(const CanonPath & path) override
+    {
+        auto blob = getBlob(path);
+
+        auto data = std::string_view((const char *) git_blob_rawcontent(blob.get()), git_blob_rawsize(blob.get()));
+
+        return std::string(data);
+    }
+
+    bool pathExists(const CanonPath & path) override
+    {
+        return path.isRoot() ? true : (bool) lookup(path);
+    }
+
+    Stat lstat(const CanonPath & path) override
+    {
+        if (path.isRoot())
+            return Stat { .type = tDirectory };
+
+        auto entry = need(path);
+
+        auto mode = git_tree_entry_filemode(entry);
+
+        if (mode == GIT_FILEMODE_TREE)
+            return Stat { .type = tDirectory };
+
+        else if (mode == GIT_FILEMODE_BLOB)
+            return Stat { .type = tRegular };
+
+        else if (mode == GIT_FILEMODE_BLOB_EXECUTABLE)
+            return Stat { .type = tRegular, .isExecutable = true };
+
+        else if (mode == GIT_FILEMODE_LINK)
+            return Stat { .type = tSymlink };
+
+        else if (mode == GIT_FILEMODE_COMMIT)
+            // Treat submodules as an empty directory.
+            return Stat { .type = tDirectory };
+
+        else
+            throw Error("file '%s' has an unsupported Git file type");
+
+    }
+
+    DirEntries readDirectory(const CanonPath & path) override
+    {
+        return std::visit(overloaded {
+            [&](Tree tree) {
+                DirEntries res;
+
+                auto count = git_tree_entrycount(tree.get());
+
+                for (size_t n = 0; n < count; ++n) {
+                    auto entry = git_tree_entry_byindex(tree.get(), n);
+                    // FIXME: add to cache
+                    res.emplace(std::string(git_tree_entry_name(entry)), DirEntry{});
+                }
+
+                return res;
+            },
+            [&](Submodule) {
+                return DirEntries();
+            }
+        }, getTree(path));
+    }
+
+    std::string readLink(const CanonPath & path) override
+    {
+        throw UnimplementedError("GitInputAccessor::readLink");
```

Comment:
Not sure why, but I hit exactly this error when trying to build system closures after switching from `github:edolstra/nix/00b746d09990a23f014b64f36e6bc09aa53cb7ec` to `github:edolstra/nix/37b4a9ec66a507484132331c8dd9d34c4c9c5778` @edolstra